### DISCUSSION TITLE
Checked Checkboxes on safari and mac os viewer not displayed

### DISCUSSION THREAD

**Discussion by sddevsd** [pymupdf/PyMuPDF#2438] [Q&A]
## Describe the bug (mandatory)
Checked Checkboxes on Safari and mac os viewer are not displayed

## To Reproduce (mandatory)
from fitz import fitz
    with fitz.open(pdffilename) as doc:
        for page in doc:
            widgets = page.widgets()
            for widget in widgets:
                widget.field_value = True
        doc.save(pdfoutputfilename)

## Expected behavior (optional)
checkbox checked on safari and mac os viewer

## Your configuration (mandatory)
 - Operating system, potentially version and bitness
 Mac Os Ventura 13.3.1
 - Python version 3.11.3
 - PyMuPDF version 1.22.3, installation method wheel 

For example, the output of `print(sys.version, "\n", sys.platform, "\n", fitz.__doc__)` would be sufficient (for the first two bullets).

3.11.3 (v3.11.3:f3909b8bc8, Apr  4 2023, 20:12:10) [Clang 13.0.0 (clang-1300.0.29.30)] 
 darwin 
 
PyMuPDF 1.22.3: Python bindings for the MuPDF 1.22.0 library.
Version date: 2023-05-10 00:00:01.
Built for Python 3.11 on darwin (64-bit).

**Comment by JorjMcKie** [comment]
Please provide all mandatory information. Currently there is no script, nor reproducer file, nor OS / PyMuPDF versions.

**Comment by sddevsd** [comment]
> Please provide all mandatory information. Currently there is no script, nor reproducer file, nor OS / PyMuPDF versions.

Yep just finished typing to update the issue.

**Comment by JorjMcKie** [comment]
Ah, ok, thank you.
You are forgetting to do `widget.update()`, don't you? And you also don't check, whether the field_type is actually what you want.

**Comment by JorjMcKie** [comment]
Cannot reproduce: works correctly on Adobe, evince (Linux), Mozilla, MS Edge browser.
Don't have a Mac and also am not using Safari, but I believe something else is causing your problem.

**Comment by sddevsd** [comment]
Hi, 

I forgot the widget.update() in the code example of the issue but not in my code.
Yes it works with chrome and Adobe but It does not works with Safari and the mac os viewer on mac os x.
Could you use a virtual machine on mac os x with virtual box ?
I can show you the issue with teamviewer on my machine if you want.

Regards,

Seb.

**Comment by JorjMcKie** [comment]
> Hi,
> 
> I forgot the widget.update() in the code example of the issue but not in my code. Yes it works with chrome and Adobe but It does not works with Safari and the mac os viewer on mac os x. Could you use a virtual machine on mac os x with virtual box ? I can show you the issue with teamviewer on my machine if you want.
> 
> Regards,
> 
> Seb.

I believe you, no worries 😎. But as all other viewers do work, including the de-facto standard Adobe, and including evince on Linux, it must be a problem specific to Mac.
Also double checked the PDF source and can confirm that the checkbox is correctly defined.
So I don't know what else I could do 🤷‍♂️

**Comment by sddevsd** [comment]
isn't there an option to make the checkboxes visible?

**Comment by JorjMcKie** [comment]
> isn't there an option to make the checkboxes visible?

It seems that is not the problem - instead I believe that these viewers do not accept the checkbox as being valid at all.

Just thought about it again - maybe the following does help. And if it does, we could add a fix for the next version:

Add the following instruction after the widget.update():

```python
widget.field_value = True
widget.update()
doc.xref_set_key(widget.xref, "V", "/Yes")
```

Please leet me know if this helps.

**Comment by sddevsd** [comment]
It's not working with doc.xref_set_key(widget.xref, "V", "/Yes")

**Comment by JorjMcKie** [comment]
No error messages or anything?
Could you try and print the PDFdefinition of the widget object? Before and after that last trick:
```python
print(doc.xref_object(widget.xref))
```

**Comment by sddevsd** [comment]
before
<<
  /FT /Btn
  /F 4
  /BS <<
    /W 1
    /S /S
  >>
  /Subtype /Widget
  /DA (0 g /Helv 12 Tf)
  /H /P
  /Border [ 0 0 0 ]
  /Rect [ 309.064 796.629 321.908 809.915 ]
  /Type /Annot
  /MK <<
    /CA (4)
  >>
  /AP <<
    /N <<
      /Off 952 0 R
      /Yes 954 0 R
    >>
  >>
  /T (INI)
  /AS /Yes
  /Ff 0
  /V (Oui)
  /TU ()


after
<<
  /FT /Btn
  /F 4
  /BS <<
    /W 1
    /S /S
  >>
  /Subtype /Widget
  /DA (0 g /Helv 12 Tf)
  /H /P
  /Border [ 0 0 0 ]
  /Rect [ 309.064 796.629 321.908 809.915 ]
  /Type /Annot
  /MK <<
    /CA (4)
  >>
  /AP <<
    /N <<
      /Off 952 0 R
      /Yes 954 0 R
    >>
  >>
  /T (INI)
  /AS /Yes
  /Ff 0
  /V /Yes
  /TU ()

**Comment by JorjMcKie** [comment]
Well, Oui = French for Yes is grotesquely wrong 😂! No idea how you managed to get this into the PDF.
The correct value in the `V` field is either "(Yes)"  or "/Yes".

So you may want to try either `doc.xref_set_key(widget.xref, "V", "/Yes")` or `doc.xref_set_key(widget.xref, "V", "(Yes)")`.

**Comment by sddevsd** [comment]
It is an official pdf from french administration xD
I am trying

**Comment by JorjMcKie** [comment]
> It is an official pdf from french administration xD I am trying

Good to know it's not your fault 😁.
But seriously: one can exaggerate insisting on the national language, right? What would happen if German admin insisted on "Ja" and "Nein", or the Spanish on "Si" and "No".
Anyway - technically, only `Yes` and `Off` are valid states of a check box. In addition, the two associated objects xref 952 and 954 are called like that and therefore must be referenced via the names assigned to them.

**Comment by sddevsd** [comment]
It's not working  either with doc.xref_set_key(widget.xref, "V", "/Yes") or doc.xref_set_key(widget.xref, "V", "(Yes)")

**Comment by sddevsd** [comment]
In the pdf form, "Oui" is the exported value for the checkbox, then I suppose you can have the values you prefer for the checkboxes. if I change it to Yes it will work ? I changed the value in the pdf form Oui to Yes and it still not displayed with safari.

**Comment by JorjMcKie** [comment]
> In the pdf form, "Oui" is the exported value for the checkbox, if I change it to Yes it will work ?

I would not bet my personal whealth on this, but yes, I believe so.

**Comment by sddevsd** [comment]
I changed the value in the pdf form Oui to Yes and it still not displayed with safari.

**Comment by sddevsd** [comment]
the checkbox is valid : in the before print of my previous export : 
/AS /Yes
/Ff 0
/V (Oui)

the /AS is /Yes before and after, only the /V have the value Oui.

/V can have the value you want with pdf adobe acrobat pro forms when you create the form.

**Comment by JorjMcKie** [comment]
Ok, I think it is clear, that we don't have an PyMuPDF issue here. So I am going to convert this to a "Discussions" item, where we can continue evaluating options...

**Comment by sddevsd** [comment]
but it's working with pdftk, why it's not working with pymupdf, I think it's an issue with pymupdf because the checkboxes are working on mac with the others softwares. I am creating a new pdf form with adobe acrobat pro, a simple form with only one checkbox and one textfield and I will do the test with this new pdf :
[Untitled.pdf](https://github.com/pymupdf/PyMuPDF/files/11616637/Untitled.pdf)

**Comment by sddevsd** [comment]
Do you have a "Need Appearances flag" option for the checkboxes ?

**Comment by sddevsd** [comment]
Then I tried with this new pdf and It's working then maybe the issue is with the pdf but I don't know what to change.

**Comment by sddevsd** [comment]
I removed the checkboxes from the pdf and added them again and now it's working, thanks a million for your help.
Have a good week.
